src/ui/components/ShowMoreCard/index.js

```diff
@@ -1,55 +1,126 @@
+/* @flow */
 /* eslint-disable react/no-danger */
-
+import invariant from 'invariant';
 import makeClassName from 'classnames';
 import * as React from 'react';
-import PropTypes from 'prop-types';
 import { compose } from 'redux';
 
 import translate from 'core/i18n/translate';
+import withUIState from 'core/withUIState';
 import { sanitizeHTML } from 'core/utils';
 import Card from 'ui/components/Card';
+import type { I18nType } from 'core/types/i18n';
 
 import './styles.scss';
 
 // This refers to height of `Card-contents` CSS class,
 // beyond which it will add read more link.
 export const MAX_HEIGHT = 150;
 
-export class ShowMoreCardBase extends React.Component {
-  static propTypes = {
-    children: PropTypes.node,
-    className: PropTypes.string,
-    header: PropTypes.node,
-    i18n: PropTypes.object.isRequired,
-  };
+type UIStateType = {|
+  showAllContent: boolean,
+  readMoreExpanded: boolean,
+|};
+
+type Props = {|
+  children: React.Element<any> | string,
+  className?: string,
+  header?: React.Element<any> | string,
+  id: string,
+|};
+
+type InternalProps = {|
+  ...Props,
+  i18n: I18nType,
+  setUIState: ($Shape<UIStateType>) => void,
+  uiState: UIStateType,
+  children: React.Element<any>,
+|};
+
+const initialUIState: UIStateType = {
+  showAllContent: true,
+  readMoreExpanded: false,
+};
+
+export class ShowMoreCardBase extends React.Component<InternalProps> {
+  contents: HTMLElement | null;
 
-  constructor(props) {
-    super(props);
-    this.state = { expanded: true };
+  componentWillReceiveProps(nextProps: InternalProps) {
+    const { children } = this.props;
+    const { children: newChildren } = nextProps;
+
+    let html =
+      children.props &&
+      children.props.dangerouslySetInnerHTML &&
+      children.props.dangerouslySetInnerHTML.__html;
+
+    let newHtml =
+      newChildren.props &&
+      newChildren.props.dangerouslySetInnerHTML &&
+      newChildren.props.dangerouslySetInnerHTML.__html;
+
+    // If it's not html, check for plain text.
+    if (!html && children && !children.props) {
+      html = children;
+    }
+
+    if (newHtml !== 'undefined' && newChildren && !newChildren.props) {
```

Comment:
Why are you checking `!html` at line 63, but here you're checking `newHtml !== 'undefined'`? Also, shouldn't it be just `undefined` and not `'undefined'`? Maybe I'm not understanding the logic of this statement.

Reply:
I see what you mean but I am sort of wondering now if we should even check this at this point. It's adding an extra level of complexity for something that is not even set up on site at this point. I feel like with this type of component - where it's pulling in variable text (which has variable length) - it's text from the user so it would always be set up with `dangerousSetInnerHtml`...

@bobsilverberg, what do you think?

Reply:
I guess we have this in the test so we should keep this and work it out..